compiler/src/iree/compiler/Codegen/Common/PropagateConstantOffsets.cpp

```diff
@@ -0,0 +1,316 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Common/Passes.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/IR/BuiltinTypeInterfaces.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_PROPAGATECONSTANTOFFSETSPASS
+#include "iree/compiler/Codegen/Common/Passes.h.inc"
+
+namespace {
+struct PropagateConstantOffsetsPass final
+    : public impl::PropagateConstantOffsetsPassBase<
+          PropagateConstantOffsetsPass> {
+  using Base::Base;
+  void runOnOperation() override;
+};
+
+/// Helper to extract constant RHS values from a defining arith op.
+template <typename OpTy>
+static std::optional<int64_t> getValueConstantRhs(Value v) {
+  auto op = v.getDefiningOp<OpTy>();
+  if (!op) {
+    return std::nullopt;
+  }
+
+  // Folders move constant operands to the RHS so no need to check both sides.
+  llvm::APInt constant;
+  if (!matchPattern(op.getRhs(), m_ConstantInt(&constant))) {
+    return std::nullopt;
+  }
+
+  return constant.getSExtValue();
+}
+
+/// Converts applies of the form:
+/// affine.apply affine_map<expr + C>
+/// to
+/// %apply = affine.apply affine_map<expr>
+/// arith.addi %apply, C
+struct ExtractConstantApplyOffset final
+    : OpRewritePattern<affine::AffineApplyOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(affine::AffineApplyOp apply,
+                                PatternRewriter &rewriter) const override {
+    AffineMap map = apply.getMap();
+    // Simplify the map to move `+ c` terms to the right most (first) expression
+    // in the tree.
+    map = simplifyAffineMap(map);
+    AffineExpr resultExpr = map.getResult(0);
+    auto addExpr = llvm::dyn_cast<AffineBinaryOpExpr>(resultExpr);
```

Comment:
Most of the `llvm::` prefixes are a result of automated refactoring after we deprecated the cast member function. We don't need them in hand-written code.
```suggestion
    auto addExpr = dyn_cast<AffineBinaryOpExpr>(resultExpr);
```